# Movie Rating Hypothesis testing

#### Grading:


- Code: 90 pts
- Markdown Documentation: 10 pts


In this assignment we are going prove of disprove a hypothesis

In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
#from IPython.display import HTML, display
#from sklearn.features_selection import chi2



# #pd.set_option('display.notebook_repr_html', False)
# pd.set_option('display.max_columns', 20)
# pd.set_option('display.max_rows', 25)

In [2]:
%matplotlib inline

##### Read and cleanup the movie data

In [3]:
def get_movie_data():
    
    unames = ['user_id','gender','age','occupation','zip']
    users = pd.read_table(os.path.join('../data','users.dat'), 
                          sep='::', header=None, names=unames, encoding='latin-1')
    
    rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
    ratings = pd.read_table(os.path.join('../data', 'ratings.dat'), 
                            sep='::', header=None, names=rnames, encoding='latin-1')
    
    mnames = ['movie_id', 'title','genres']
    movies = pd.read_table(os.path.join('../data', 'movies.dat'), 
                           sep='::', header=None, names=mnames, encoding='latin-1')

    return users, ratings, movies

users, ratings, movies = get_movie_data()

tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')
tmp.apply(lambda x:x[0] if len(x) > 0 else None)
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)

movies['year'] = tmp[1]
movies['short_title'] = tmp[0]

print(movies.head())

/Users/vijaykrishna/opt/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py:311: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return func(*args, **kwargs)


   movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II  


### 1. Prove or disprove the following Null Hypothesis:

#### H0: Different age segments of users rate the different genres of movies similarly

- Age segments are (0-20), (21-30), (31-50), (50 and above),
- Genres - Drama, Comedy, Action, Romance, Adventure
- Rating 4 or above is a positive, else negative (used for binary votes)

#### Solution

# I Implemented the chi2 test of independence in this assignment

In [4]:
print(users.head())

   user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455


In [5]:
print(ratings.head())

   user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291


In [6]:
print(movies.head())

   movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II  


In [7]:
ratings_users = pd.merge(users,ratings, left_on='user_id', right_on='user_id')
ratings_users_movies = pd.merge(movies,ratings_users, left_on='movie_id' , right_on = 'movie_id')

ratings_users_movies

movie_id                  title                       genres  year  \
0               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
1               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
2               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
3               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
4               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
...           ...                    ...                          ...   ...   
1000204      3952  Contender, The (2000)               Drama|Thriller  2000   
1000205      3952  Contender, The (2000)               Drama|Thriller  2000   
1000206      3952  Contender, The (2000)               Drama|Thriller  2000   
1000207      3952  Contender, The (2000)               Drama|Thriller  2000   
1000208      3952  Contender, The (2000)               Drama|Thriller  2000   

            short_title  user_id gender  age  occupation    zip  rating  \
0             Toy Story        1      F    1          10  48067       5   
1             Toy Story        6      F   50           9  55117       4   
2             Toy Story        8      M   25          12  11413       4   
3             Toy Story        9      M   25          17  61614       5   
4             Toy Story       10      F   35           1  95370       5   
...                 ...      ...    ...  ...         ...    ...     ...   
1000204  Contender, The     5812      F   25           7  92120       4   
1000205  Contender, The     5831      M   25           1  92120       3   
1000206  Contender, The     5837      M   25           7  60607       4   
1000207  Contender, The     5927      M   35          14  10003       1   
1000208  Contender, The     5998      M   18           4  61820       4   

          timestamp  
0         978824268  
1         978237008  
2         978233496  
3         978225952  
4         978226474  
...             ...  
1000204   992072099  
1000205   986223125  
1000206  1011902656  
1000207   979852537  
1000208  1001781044  

[1000209 rows x 12 columns]

- codification of ratings label
- Rating 4 or above is a positive, else negative (used for binary votes)
- ratings of 4 and 5 will be '1'(positive) and lower ratings are '0' (negitive)

In [8]:
cat = []
for i in range (0, 1):
    for value in ratings_users_movies["rating"]:
        if value >= 4 :
            cat.append("1")
        elif value < 4:
            cat.append("0")



In [9]:
ratings_users_movies["ratings_label"] = cat


In [10]:
ratings_users_movies.head(10)

movie_id             title                       genres  year short_title  \
0         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
1         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
2         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
3         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
4         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
5         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
6         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
7         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
8         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   
9         1  Toy Story (1995)  Animation|Children's|Comedy  1995   Toy Story   

   user_id gender  age  occupation    zip  rating  timestamp ratings_label  
0        1      F    1          10  48067       5  978824268             1  
1        6      F   50           9  55117       4  978237008             1  
2        8      M   25          12  11413       4  978233496             1  
3        9      M   25          17  61614       5  978225952             1  
4       10      F   35           1  95370       5  978226474             1  
5       18      F   18           3  95825       4  978154768             1  
6       19      M    1          10  48073       5  978555994             1  
7       21      M   18          16  99353       3  978139347             0  
8       23      M   35           0  90049       4  978463614             1  
9       26      M   25           7  23112       3  978130703             0

- **now codification of age into 4 groups**

In [11]:
group = []
for i in range (0, 1):
    for value in ratings_users_movies["age"]:
        if value >= 0 and value <= 20:
            group.append("1")
        elif value >= 21 and value <= 30:
            group.append("2")
        elif value >= 31 and value <= 50 :
            group.append("3")
        elif value > 50 :
            group.append("4")

        
    
#df_n["cat"] = cat   

ratings_users_movies["age_label"] = group

- Age segments are (0-20), (21-30), (31-50), (50 and above),
- (0-20) is 1
- (21 - 30) is 2
- (31 - 50) is 3
- (51 and above) is 4


In [12]:
ratings_users_movies

movie_id                  title                       genres  year  \
0               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
1               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
2               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
3               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
4               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
...           ...                    ...                          ...   ...   
1000204      3952  Contender, The (2000)               Drama|Thriller  2000   
1000205      3952  Contender, The (2000)               Drama|Thriller  2000   
1000206      3952  Contender, The (2000)               Drama|Thriller  2000   
1000207      3952  Contender, The (2000)               Drama|Thriller  2000   
1000208      3952  Contender, The (2000)               Drama|Thriller  2000   

            short_title  user_id gender  age  occupation    zip  rating  \
0             Toy Story        1      F    1          10  48067       5   
1             Toy Story        6      F   50           9  55117       4   
2             Toy Story        8      M   25          12  11413       4   
3             Toy Story        9      M   25          17  61614       5   
4             Toy Story       10      F   35           1  95370       5   
...                 ...      ...    ...  ...         ...    ...     ...   
1000204  Contender, The     5812      F   25           7  92120       4   
1000205  Contender, The     5831      M   25           1  92120       3   
1000206  Contender, The     5837      M   25           7  60607       4   
1000207  Contender, The     5927      M   35          14  10003       1   
1000208  Contender, The     5998      M   18           4  61820       4   

          timestamp ratings_label age_label  
0         978824268             1         1  
1         978237008             1         3  
2         978233496             1         2  
3         978225952             1         2  
4         978226474             1         3  
...             ...           ...       ...  
1000204   992072099             1         2  
1000205   986223125             0         2  
1000206  1011902656             1         2  
1000207   979852537             0         3  
1000208  1001781044             1         1  

[1000209 rows x 14 columns]

- Extracting required genres from the dataset
- **Genres - Drama, Comedy, Action, Romance, Adventure**

In [13]:
genres = ratings_users_movies.loc[ratings_users_movies['genres'].str.contains('Drama'and'Action'and'Adventure'and'Comedy'and'Romance')]

#genres = ratings_users_movies.loc[ratings_users_movies.genres == (['Drama'|'Action'|'Adventure'|'Comedy'|'Romance'])]

#genres = ratings_users_movies.loc[ratings_users_movies['genres'].isin(['Drama', 'Action', 'Adventure', 'Comedy', 'Romance'])]




In [14]:
genres

movie_id                    title          genres  year  \
2778           3  Grumpier Old Men (1995)  Comedy|Romance  1995   
2779           3  Grumpier Old Men (1995)  Comedy|Romance  1995   
2780           3  Grumpier Old Men (1995)  Comedy|Romance  1995   
2781           3  Grumpier Old Men (1995)  Comedy|Romance  1995   
2782           3  Grumpier Old Men (1995)  Comedy|Romance  1995   
...          ...                      ...             ...   ...   
993956      3909      Woman on Top (2000)  Comedy|Romance  2000   
993957      3909      Woman on Top (2000)  Comedy|Romance  2000   
993958      3909      Woman on Top (2000)  Comedy|Romance  2000   
993959      3909      Woman on Top (2000)  Comedy|Romance  2000   
993960      3909      Woman on Top (2000)  Comedy|Romance  2000   

             short_title  user_id gender  age  occupation    zip  rating  \
2778    Grumpier Old Men       26      M   25           7  23112       2   
2779    Grumpier Old Men       45      F   45          16  94110       2   
2780    Grumpier Old Men       62      F   35           3  98105       3   
2781    Grumpier Old Men      137      F   45           6  78758       5   
2782    Grumpier Old Men      153      M    1          10  51537       5   
...                  ...      ...    ...  ...         ...    ...     ...   
993956      Woman on Top     5405      F   50           7  12866       2   
993957      Woman on Top     5539      F   25          20  91403       1   
993958      Woman on Top     5874      M   25           4  01002       3   
993959      Woman on Top     5995      F   35           1  14618       3   
993960      Woman on Top     6016      M   45           1  37209       2   

         timestamp ratings_label age_label  
2778     978140022             0         2  
2779     977989159             0         3  
2780     977904968             0         3  
2781    1000869657             1         3  
2782     977437594             1         1  
...            ...           ...       ...  
993956   994011380             0         3  
993957   973528981             0         2  
993958   975280235             0         2  
993959   995385509             0         3  
993960   994455188             0         3  

[147523 rows x 14 columns]

In [15]:
final = genres[genres['ratings_label']=='1']
# extraxting the dataframe with ratings positive ratings only

- pd.crosstab method is used to compute a simple cross-tabulation of two (or more) factors. 
- By default, it computes a frequency table of the factors unless an array of values and an aggregation function are passed

In [16]:
table = pd.crosstab(final.genres,final.age_label, margins=True)

In [17]:
table
# all the positive ratings are only considered
# for all the genres four age segments of user are extracted

age_label                                        1      2      3     4    All
genres                                                                       
Action|Adventure|Comedy|Romance                827   1575   1293   113   3808
Action|Adventure|Drama|Romance                  32     46     42     2    122
Action|Adventure|Romance                       154    186    224    19    583
Action|Adventure|Romance|Sci-Fi|War            542    856    653    76   2127
Action|Adventure|Romance|Thriller              105    189    204    20    518
Action|Adventure|Romance|War                    60    204    543    87    894
Action|Comedy|Romance|Thriller                  15     39     46     2    102
Action|Crime|Romance                           503    909    492    58   1962
Action|Drama|Fantasy|Romance                    52    224    208     6    490
Action|Drama|Mystery|Romance|Thriller            4     52    128    11    195
Action|Drama|Romance                            79    144    143    23    389
Action|Drama|Romance|Thriller                   27     86     74     6    193
Action|Mystery|Romance|Thriller                110    143    135    12    400
Action|Romance                                 219    359    254    21    853
Action|Romance|Sci-Fi                           47    111     85    13    256
Action|Romance|Thriller                        618   1003    696    56   2373
Action|Romance|War                             141    232    244    28    645
Adventure|Children's|Comedy|Fantasy|Romance     16      5     10     0     31
Adventure|Children's|Drama|Romance              11      3     11     2     27
Adventure|Children's|Romance                    29     33     44     1    107
Adventure|Comedy|Romance                        26     40     60     4    130
Adventure|Drama|Romance                         11     24     39     7     81
Adventure|Drama|Romance|Sci-Fi                  29    151    239    25    444
Adventure|Fantasy|Romance                       98    219    152    10    479
Adventure|Musical|Romance                        3      6      9     0     18
Adventure|Romance                               32     74     47     3    156
Adventure|Romance|Sci-Fi                        44     50     84     6    184
Animation|Children's|Comedy|Musical|Romance    321    457    433    37   1248
Animation|Children's|Comedy|Romance             22     13     15     1     51
Animation|Children's|Musical|Romance            46     27     48     4    125
Comedy|Drama|Romance                          1051   2721   2590   321   6683
Comedy|Fantasy|Romance                          81    204    246    16    547
Comedy|Fantasy|Romance|Sci-Fi                  102    124     89     5    320
Comedy|Musical|Romance                         177    390    478    83   1128
Comedy|Mystery|Romance                           8      8      8     2     26
Comedy|Mystery|Romance|Thriller                 39    112    222    36    409
Comedy|Romance                                4762   8751   8795  1045  23353
Comedy|Romance|Sci-Fi                            3     11     11     1     26
Comedy|Romance|Thriller                        288    491    448    56   1283
Comedy|Romance|War                             413    616    549    90   1668
Crime|Drama|Romance                             25     83    164    30    302
Crime|Drama|Romance|Thriller                   251    536    314    25   1126
Drama|Fantasy|Romance|Thriller                  55    159    109     9    332
Drama|Mystery|Romance                           36    114     90    12    252
Drama|Romance                                 2991   6469   6644   970  17074
Drama|Romance|Sci-Fi                            10     19     16     2     47
Drama|Romance|Thriller                          83    370    447    47    947
Drama|Romance|War                              596   1493   2217   410   4716
Drama|Romance|War|Western                       69    153    117    21    360
Drama|Romance|Western 

In [18]:
table.shape

(60, 5)

In [19]:
newdf = table.drop("All", axis='columns')
newdf = newdf.drop('All', axis='index')

# from the table dropping the unwanted rows and columns

In [20]:
newdf.shape

(59, 4)

In [21]:


newdf.columns = ["age(0-20)","age(21-30)","age(31-50)","age(51 and above)"]

# reassigning the column names for the table below

In [22]:
newdf

age(0-20)  age(21-30)  \
genres                                                               
Action|Adventure|Comedy|Romance                    827        1575   
Action|Adventure|Drama|Romance                      32          46   
Action|Adventure|Romance                           154         186   
Action|Adventure|Romance|Sci-Fi|War                542         856   
Action|Adventure|Romance|Thriller                  105         189   
Action|Adventure|Romance|War                        60         204   
Action|Comedy|Romance|Thriller                      15          39   
Action|Crime|Romance                               503         909   
Action|Drama|Fantasy|Romance                        52         224   
Action|Drama|Mystery|Romance|Thriller                4          52   
Action|Drama|Romance                                79         144   
Action|Drama|Romance|Thriller                       27          86   
Action|Mystery|Romance|Thriller                    110         143   
Action|Romance                                     219         359   
Action|Romance|Sci-Fi                               47         111   
Action|Romance|Thriller                            618        1003   
Action|Romance|War                                 141         232   
Adventure|Children's|Comedy|Fantasy|Romance         16           5   
Adventure|Children's|Drama|Romance                  11           3   
Adventure|Children's|Romance                        29          33   
Adventure|Comedy|Romance                            26          40   
Adventure|Drama|Romance                             11          24   
Adventure|Drama|Romance|Sci-Fi                      29         151   
Adventure|Fantasy|Romance                           98         219   
Adventure|Musical|Romance                            3           6   
Adventure|Romance                                   32          74   
Adventure|Romance|Sci-Fi                            44          50   
Animation|Children's|Comedy|Musical|Romance        321         457   
Animation|Children's|Comedy|Romance                 22          13   
Animation|Children's|Musical|Romance                46          27   
Comedy|Drama|Romance                              1051        2721   
Comedy|Fantasy|Romance                              81         204   
Comedy|Fantasy|Romance|Sci-Fi                      102         124   
Comedy|Musical|Romance                             177         390   
Comedy|Mystery|Romance                               8           8   
Comedy|Mystery|Romance|Thriller                     39         112   
Comedy|Romance                                    4762        8751   
Comedy|Romance|Sci-Fi                                3          11   
Comedy|Romance|Thriller                            288         491   
Comedy|Romance|War                                 413         616   
Crime|Drama|Romance                                 25          83   
Crime|Drama|Romance|Thriller                       251         536   
Drama|Fantasy|Romance|Thriller                      55         159   
Drama|Mystery|Romance                               36         114   
Drama|Romance                                     2991        6469   
Drama|Romance|Sci-Fi                                10          19   
Drama|Romance|Thriller                              83         370   
Drama|Romance|War                                  596        1493   
Drama|Romance|War|Western                           69         153   
Drama|Romance|Western                                1           2   
Film-Noir|Romance|Thriller                          45         127   
Horror|Romance                                     303         418   
Musical|Romance                                    360         627   
Musical|Romance|War                                 18          24   
Mystery|Romance|Thriller                            69         241   
Romance                                            175 

In [23]:
stats.chi2_contingency(observed= newdf)

(2824.940604078445,
 0.0,
 174,
 array([[7.26601621e+02, 1.43529875e+03, 1.45921594e+03, 1.86883683e+02],
        [2.32787284e+01, 4.59838361e+01, 4.67500906e+01, 5.98734488e+00],
        [1.11241792e+02, 2.19742430e+02, 2.23404122e+02, 2.86116563e+01],
        [4.05851273e+02, 8.01701798e+02, 8.15061007e+02, 1.04385923e+02],
        [9.88391911e+01, 1.95242845e+02, 1.98496287e+02, 2.54216774e+01],
        [1.70583469e+02, 3.36963520e+02, 3.42578533e+02, 4.38744781e+01],
        [1.94625434e+01, 3.84455023e+01, 3.90861414e+01, 5.00581293e+00],
        [3.74367747e+02, 7.39510544e+02, 7.51833425e+02, 9.62882841e+01],
        [9.34965321e+01, 1.84689178e+02, 1.87766758e+02, 2.40475327e+01],
        [3.72078036e+01, 7.34987544e+01, 7.47235055e+01, 9.56993649e+00],
        [7.42247980e+01, 1.46620592e+02, 1.49063814e+02, 1.90907964e+01],
        [3.68261851e+01, 7.27449210e+01, 7.39571106e+01, 9.47178330e+00],
        [7.63236997e+01, 1.50766676e+02, 1.53278986e+02, 1.96306390e+01],
      

- 'p' value is 0
- chi2 statistic value = 2824.9
- Degrees of Freedom = 174
- expected array - same shape as the observed data table based on marginal sums of the table

In [24]:
crit = stats.chi2.ppf(q = 0.95, # Find the critical value for 95% confidence
                      df = 174)

In [25]:
print('Critical value is:\n', crit)
# critical value is given below

Critical value is:
 205.77862677980613


# we reject the null hypothesis as the p-value obtained is '0'

In [ ]:
#